The `navigate-to` directive

andypaicu opened the pull request:
There has been talk at TPAC about this and there is some generally positive feedback around [this explainer](https://docs.google.com/document/d/1eMfw7sSIPtPPs9T3K2C8SfDi3Q7OXRTrRDdkGOLb19M/edit?ts=5a5e00ef) I sent earlier.

WDYT?

Also who else should be added?


<!--
    This comment and the below content is programatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/andypaicu/webappsec-csp/pull/290.html" title="Last updated on Mar 5, 2018, 8:34 PM GMT (2be467a)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/webappsec-csp/290/e333f87...andypaicu:2be467a.html" title="Last updated on Mar 5, 2018, 8:34 PM GMT (2be467a)">Diff</a>

annevk commented:
Bikeshed: "navigate-to". Justification: it's "connect-src", not "connection-src". (I might even go as far as just "navigate" as I'm not sure what the suffix adds.)

andypaicu commented:
I can see `navigate-to` seems reasonable. I would like to keep the `to` suffix as I think it is somewhat clearer that we are specifying destinations and not perhaps something else about navigations.

I have some concern and I'm not sure what happens when the initiator is not a browsing context (for example a url typed in the box) but I want to ensure that this will never block those navigations.

mikewest commented on index.src.html:
```diff
@@ -548,13 +544,14 @@ <h3 id="framework-directives">Directives</h3>
   7.  A <dfn for="directive" export>navigation response check</dfn>, which takes a
-      <a for="/">request</a>, a <a>response</a> and two <a>browsing contexts</a> as
-      arguments, and is executed during [[#should-block-navigation-response]].
+      <a for="/">request</a>, a <a>response</a>, two <a>browsing contexts</a>,
```
This also takes a "navigation type" string (either `form-submission` or `other`).

mikewest commented on index.src.html:
```diff
@@ -548,13 +544,14 @@ <h3 id="framework-directives">Directives</h3>
-      <a for="/">request</a>, type string, and two <a>browsing contexts</a> as arguments, and
-      is executed during [[#should-block-navigation-request]]. It returns
-      "`Allowed`" unless otherwise specified.
+      <a for="/">request</a>, `navigation type` string, and two <a>browsing contexts</a>,
```
Nit: Drop the backticks: "navigation type string" seems fine without them. Same below for "check type string".

mikewest commented on index.src.html:
```diff
@@ -1386,7 +1383,31 @@ <h4 id="should-block-navigation-response" algorithm>
             1.  If |directive|'s <a for="directive">navigation response check</a>
                 returns "`Allowed`" when executed upon |navigation request|, |type|,
-                |navigation response|, |source|, and |target|, skip to the next
+                |navigation response|, |source|, |target|, and "`target`" skip to the next
```
I think it's worth adding a note under step 2 above explaining that this piece of the algorithm is useful for directives that allow the response itself to assert something. Perhaps along the lines of:

> Some directives, like [=frame-ancestors=], allow a policy delivered with |navigation response| to
> act upon the navigation. To support these directives, we walk through the response's CSP list
> and allow its directives to process their [=directive/navigation response checks=]

Under the new step 3, you might follow that up with:

> Other directives, like [=navigate-to=], allow the policies associated with the navigation's |source|
> to act upon the navigation. To support these directives, we walk through the relevant document's
> CSP list, and allow its directives to process their [=directive/navigation response checks=].

That framing might also suggest switching the terminology here from "source"/"target" to something like "response"/"source".

mikewest commented on index.src.html:
```diff
@@ -1386,7 +1383,31 @@ <h4 id="should-block-navigation-response" algorithm>
+                |directive|'s <a for="directive">name</a>.
+
+                Note: We use `null` for the global object, as no global exists:
+                we haven't processed the navigation to create a Document yet.
```
This makes sense for the target checks. I wonder if we ought to use |source|'s global for the source checks. That might not actually be possible, give the way browsers implement navigation, of course. Chrome might have thrown away the initial context if we're processing a redirect at this point...

mikewest commented on index.src.html:
```diff
@@ -1386,7 +1383,31 @@ <h4 id="should-block-navigation-response" algorithm>
+                we haven't processed the navigation to create a Document yet.
+
+            3.  Set |violation|'s <a for="violation">resource</a> to |navigation
+                response|'s <a for="response">URL</a>.
```
This also makes sense for the target checks, but I think it might be wrong for the source checks. We'd want the initial URL of the request that began the navigation. That is, given the chain `A => B => C`, if `C` is a violation, we should return `A` in order to avoid leaking redirect targets. Same logic as the note in step 2 of https://w3c.github.io/webappsec-csp/#create-violation-for-request: something like "|navigation request|'s [=request/url=]" should do it for that section, I think.

mikewest commented on index.src.html:
```diff
@@ -3324,17 +3345,23 @@ <h5 algorithm id="frame-ancestors-navigation-response">
+    2.  If |check type| is "`source`", return "`Allowed`".
+
+        Note: The 'frame-ancestors' <a>directive</a> is relevant only on the
+        |target| <a>browsing context</a> and it has no impact on the |source|
```
It's not really relevant to the target browsing context, but to the policies delivered with the response that we're processing.

mikewest replied:
Still relevant?

mikewest commented on index.src.html:
```diff
@@ -3376,47 +3403,105 @@ <h5 id="frame-ancestors-and-frame-options">
-    a document can navigate by any means (<{a}>, <{form}>, `window.location`,
-    `window.open`, etc.). The directive's syntax is described by the following ABNF grammar:
+  The <dfn export>navigation-to</dfn> directive restricts the {{URL}}s to which
+  a <a>document</a> can initiate navigations by any means (<{a}>, <{form}>, `window.location`,
```
I wonder if we should structure this as sitting under `form-action`? That is, given the policy `navigation-to a b c; form-action d`, would we allow or deny a form submission to `d`? It seems like we'd want to allow it.

mikewest commented on index.src.html:
```diff
@@ -3376,47 +3403,105 @@ <h5 id="frame-ancestors-and-frame-options">
+    `navigation-to` Navigation Response Check
+  </h5>
+
+  Given a <a for="/">request</a> (|request|), a <a>response</a> (|navigation response|)
```
Nit: This gets the navigation type string as well.

mikewest commented on index.src.html:
```diff
@@ -3376,47 +3403,105 @@ <h5 id="frame-ancestors-and-frame-options">
+    2.  If |check type| is "`target`", return "`Allowed`".
+
+        Note: The 'navigation-to' <a>directive</a> is relevant only on the
+        |source| <a>browsing context</a> and it has no impact on the |target|
```
Nit: "relevant to", not "relevant on".

mikewest commented on index.src.html:
```diff
@@ -3376,47 +3403,105 @@ <h5 id="frame-ancestors-and-frame-options">
+
+    2.  If this directive's <a for="directive">value</a> does not contain a <a>source
+        expression</a> that is an <a>ASCII case-insensitive</a> match for
+        the "<a grammar>`'only-check-last-endpoint'`</a>"
```
How about `'allow-redirects'`? Or, in keeping with the rest of everything in CSP, `'unsafe-allow-redirects'`?

andypaicu commented:
Thank you @annevk and @mikewest for the feedback. I have addressed all of it (hopefully):

renamed `navigation-to` to `navigate-to`
form submissions will be ignored by `navigate-to` if `form-action` is present
notes on why we need to check policies from the source context when we get a response as well as the response's own policies
using the source's global object if possible (with a note for the caveat that it might not be)
various small typos, missed parameters and so on
renamed the flag to 'unsafe-allow-redirects'



I need some help with making sure that this navigation-to check does not affect any out-of-document navigations (eg typing in the omnibox). I assume the source context won't have any policies if it's not a document that initiates the navigation but I could use some help to make sure that is the case.

mikewest commented on index.src.html:
```diff
@@ -544,15 +544,15 @@ <h3 id="framework-directives">Directives</h3>
-      executed during [[#should-block-navigation-response]].
-      It returns "`Allowed`" unless otherwise specified.
+      <a for="/">request</a>, a navigation type string, a <a>response</a>, two
+      <a>browsing contexts</a>, a check type string ("`source`" or "`response`"),
```
Listing the values here is nice! Could you do it for the navigation type strings as well (```"`form-submission`" or "`other`"```)?

mikewest commented on index.src.html:
```diff
@@ -608,7 +608,7 @@ <h4 id="framework-directive-source-list">Source Lists</h4>
     <dfn>keyword-source</dfn> = "<dfn>'self'</dfn>" / "<dfn>'unsafe-inline'</dfn>" / "<dfn>'unsafe-eval'</dfn>"
                      / "<dfn>'strict-dynamic'</dfn>" / "<dfn>'unsafe-hashed-attributes'</dfn>" /
-                     / "<dfn>'report-sample'</dfn>" / "<dfn>'only-check-last-endpoint'</dfn>"
+                     / "<dfn>'report-sample'</dfn>" / "<dfn>'unsafe-allow-redirects'</dfn>"
```
Would you mind adding something like:

```
ISSUE: Bikeshed `unsafe-allow-redirects`.
```

Shouldn't block this patch, but we should think about it a bit as a group.

mikewest commented on index.src.html:
```diff
@@ -1379,12 +1379,15 @@ <h4 id="should-block-navigation-response" algorithm>
         <a for="response">CSP list</a>:
 
+        Note: Some directives (like <a>frame-ancestors</a>) allow a |response|'s
+        <a>Content Security Policy</a> to act on the navigation.
```
I liked the rest of the explanation I suggested. :) "To support these directives, we walk through the response's CSP list and allow its directives to process their [=directive/navigation response checks=]". Too much explanation?

andypaicu replied:
It felt like I was using words to describe the steps of the algorithm without adding extra information.

mikewest replied:
*shrug* Ok. I'm fine with clarifying repetition, but if you think it's too much, I'll not push on it.

mikewest commented on index.src.html:
```diff
@@ -1403,22 +1406,27 @@ <h4 id="should-block-navigation-response" algorithm>
     3.  For each |policy| in |source|'s <a>active document</a>'s <a for="Document">CSP List</a>:
 
+        Note: Some directives (like <a>navigate-to</a>) need the |response| before
+        acting on the navigation.
```
I saw the interesting bit here as acting on the policies associated with `|source|` rather than `|response|` (or `|target|`, for that matter).

mikewest commented on index.src.html:
```diff
@@ -1403,22 +1406,27 @@ <h4 id="should-block-navigation-response" algorithm>
+                Note: Depending on the browser's implementation it might not be
+                possible to obtain the |source|'s <a>relevant global object</a>
+                at this point in time. If that is the case `null` should be used
+                instead.
```
I'd drop this note. We're already relying on `|source|` active document's policy above. If that exists, then from a spec perspective, the global should exist as well. If we need to do some weird monkey patching in Chrome to store the details we need to generate the report, so be it.

annevk commented:
The eventual commit message should mention that this fixes #125.

andypaicu commented:
Once this makes it in, I will update the tests in that CR of course.

andypaicu commented:
I have separated the tests into their own PR independent of the chrome implementation patch.

https://github.com/w3c/web-platform-tests/pull/9644
